Serialize the newline_list to avoid recomputing it again later

eregon opened the pull request:
* Fixes https://github.com/ruby/prism/issues/2380

Related: https://github.com/ruby/prism/issues/2402

eregon commented:
Should be green and ready for review now.

enebo reviewed (changes requested):
Something bothers me that we are putting Source into ParseResult.  I feel like we should consider only putting LineOffsets into it.  Our end goal is to not have any source by the time we get our serialized source back.  We just have not made it to that point yet.

If this is true then Source is nothing more than startLine and source bytes[].  Do we actually need startLine or could this be boiled into the linecalcs as well?  I did not read that but I wonder if we ever need that value after parsing is over.  If so then Source can go away as well.

enebo commented:
@eregon Also simple error in build-java failing in CI but you obviously will fix that.

eregon commented:
@enebo I think it's a lot cleaner and less error-prone this way, and it's also exactly how it works in the Ruby API (how `Prism.parse` takes a String of code and returns a ParseResult which includes the Prism::Source).
The main concern there is that if someone creates a Source object they would end up with invalid lineOffsets, unless the same object is also passed to Loader.
If the Source object is created by Loader it's not possible to have this issue.
Also creating the Source object in the caller felt awkward, I'd rather Loader takes just a `(byte[] serialize, byte[] sourceBytes)`.

> Do we actually need startLine or could this be boiled into the linecalcs as well?

It cannot be stored in the lineOffsets (which are offsets in the byte[]/char* source, so 0-based), it needs to be a separate field.

Regarding once we don't need the source bytes anymore for Loader, I think that's fine, then one would be able to just pass `null` for the `sourceBytes` (or have an overload doing that). And then the Source object would only contain startLine + lineOffsets, which seems all good (well except the name maybe, but its main purpose is to provide *source* line & column info).
In fact the only usage for `bytes` in the Source Java class currently is `bytes.length`, which we could do another way or avoid needing entirely in `findLine()`.

It is also something this PR helps significantly, since the lineOffsets now come from the serialized form and there is no longer a need to use the `sourceBytes` to find them.

If it's problematic later for some unforeseen we can of course always rework this, it's pretty easy to change, b836f031cad936826b5a5e6fcb1cb071cd665ded is a pretty small commit.

Are you OK with that?
If not I can remove that commit, but IMO it is significantly cleaner this way and also more convenient to call Loader. I was confused for a while how startLine was set before, if the Loader creates the Source there is no hidden magic and it's obvious.

enebo commented:
@eregon yeah I forgot we are just getting an index as well so I did not think through the startLine comment.

Will you be keeping the source?  I don't see anyone doing this with location fields removed but for FFI consumers who have all the location data they probably do want the source (although they do already have it). 

I am ok though with landing this.  We can argue out how we deal with source bytes once we actually don't need it for anything.  I am isolating this code outside the JRuby codebase so these APIs can change as much as they want and it won't require a new JRuby release.  Just a new gem release containing a newly made provider.

enebo commented:
@eregon I am only approving Java changes as I did not look very closely at the rest.

eregon commented:
> Will you be keeping the source?

Until TruffleRuby can cache the serialized form on disk (TBD where/how, not for the 24.0 release) it needs to read the source as a byte[] anyway to pass it to parseAndSerialize().
(and of course currently the source bytes are needed for Loader so it's not yet possible to skip reading the source file)

>  I am isolating this code outside the JRuby codebase so these APIs can change as much as they want and it won't require a new JRuby release. Just a new gem release containing a newly made provider.

Don't you vendor a specific version/commit of Prism in JRuby when using it as the parser for Ruby code?
That's what TruffleRuby and CRuby do. Then it doesn't matter which version(s) of Prism are installed as gems, it's independent.

Or do you also use the Java Loader when Prism is used as a gem? I wouldn't expect that since what's tested in CI on JRuby is the Ruby API using FFI (so not using the Java Loader).

enebo commented:
> Don't you vendor a specific version/commit of Prism in JRuby when using it as the parser for Ruby code? That's what TruffleRuby and CRuby do. Then it doesn't matter which version(s) of Prism are installed as gems, it's independent.
> 
> Or do you also use the Java Loader when Prism is used as a gem? I wouldn't expect that since what's tested in CI on JRuby is the Ruby API using FFI (so not using the Java Loader).

I am isolating the Java part as an artifact and it is included into a Ruby gem which is effectively a "barely" fork of prism gem that has been renamed.  That gem includes the artifact and will compile the source with the ENVs we want when the gem is installed (later we will probably pre-compile to remove compile requirement on main platforms).

The rationale for doing this is if there is a CVE or just a good fix there is a high likelihood I only have to do a gem release and not a full JRuby release.

eregon commented:
I see, makes sense. Could be worth documenting at https://github.com/ruby/prism/blob/main/docs/build_system.md#building-prism-as-part-of-jruby
We cannot do this for TruffleRuby because it needs to work for `--native` mode and that needs to know all Java code ahead-of-time.

eregon commented:
@kddnewton Could you merge this?

kddnewton commented on templates/src/serialize.c.erb:
```diff
@@ -128,6 +128,17 @@ pm_serialize_node(pm_parser_t *parser, pm_node_t *node, pm_buffer_t *buffer) {
     }
 }
 
+void
```
```suggestion
static void
```

kddnewton commented on templates/ext/prism/api_node.c.erb:
```diff
@@ -36,18 +36,26 @@ pm_string_new(pm_string_t *string, rb_encoding *encoding) {
 
+void
+pm_source_init(VALUE source, pm_parser_t *parser) {
+    rb_funcall(source, rb_intern("start_line="), 1, LONG2NUM(parser->start_line));
```
I'd prefer not to go through the ceremony of making two method calls in this function. We already have the start line when the source is created, and when we create the array for the offsets we can keep that reference around and push directly into it. That'll avoid the frame and everything else that comes from `rb_funcall`.

eregon replied:
> We already have the start line when the source is created,

This is handled here to be consistent with how Loader.java and FFI are doing things on master (and there it's not known before what is the start line, we know it after deserializing it).
https://github.com/ruby/prism/blob/18ada3120231e5b7faa821b0710f5c19f79fa11a/templates/lib/prism/serialize.rb.erb#L81-L83

I think it's valuable to be consistent here to avoid errors and help understanding how it all works.

> when we create the array for the offsets we can keep that reference around and push directly into it

If we do that then we cannot use `rb_ary_new_capa(parser->newline_list.size)` and so it will occur extra resizes (so extra allocations and memory copies).

> I'd prefer not to go through the ceremony of making two method calls in this function.

I can change them to `rb_iv_set()` if you want. But that would be inconsistent with how FFI does things.

kddnewton replied:
In that case you can call `rb_ary_resize`. I don't want us to be setting instance variables later if we don't have to.

eregon replied:
That's not available for the corresponding piece of code in `serialize.rb.erb`.

I want `@start_line` and `@offsets` to be handled consistently for all languages and all places setting it (for readability, maintainability and avoiding bugs), and for `@start_line` it needs to be assigned (e.g. for `Parser.load`), there is no alternative.
I think that is reasonable and more important than avoiding to assign ivars (for this case).

BTW the Java code deserializing the offsets naturally returns a new array, so it needs to assign, no alternative possible there.

I do agree setting instance variables later is not ideal in general, but there is no way to avoid mutating the Source object (thinking of it as an object graph, which includes the offsets) after it's built here, because the offsets are only available after parsing. I think this mutation is harmless since the user cannot access the Source object in an incomplete state.
I refactored the Java API a bit specifically to have the same property of no incomplete Source object accessible by the user.

kddnewton replied:
As an alternative, if you cannot have it be inconsistent, please make the Java side work like the CRuby side.

eregon replied:
Done.
To my surprise, it is indeed measurably faster for `Prism.parse("1 + 2")` (see commit message for the results).
I would have expected these 2 calls to rb_iv_set()/rb_funcall() to be insignificant (and they might be for real files with more core/many lines, but not for "1 + 2").
Also interesting is that `rb_iv_set()` is quite slow on CRuby, to the point that `rb_funcall()` (for an attr_writer) is twice as fast: https://gist.github.com/eregon/a6404a71151b30021ca081904f7149a0

OTOH for serialize.rb, the fastest would be public `attr_writer`.

kddnewton commented on templates/lib/prism/serialize.rb.erb:
```diff
@@ -82,6 +82,11 @@ module Prism
       end
 
+      def load_line_offsets
+        offsets = load_varuint.times.map { load_varuint }
+        source.offsets = offsets
```
```suggestion
        source.offsets.concat(load_varuint.times.map { load_varuint })
```

eregon replied:
`concat` makes an extra copy (memcpy/arraycopy of all elements) so that's worse performance-wise.
I'll remove the extra local variable though.

kddnewton replied:
In that case can you 

```ruby
load_varuint.times do
  source.offsets << load_varuint
end
```

eregon replied:
But that will cause extra resizes, same as https://github.com/ruby/prism/pull/2381#discussion_r1488204019

kddnewton commented on lib/prism/parse_result.rb:
```diff
@@ -12,15 +12,13 @@ class Source
 
     # The list of newline byte offsets in the source code.
-    attr_reader :offsets
+    attr_accessor :offsets
```
I'd rather not have this be `attr_accessor`, it's not meant to be set later.

eregon replied:
I changed both start_line and offsets to be `attr_reader`.
And then used `#instance_variable_set/rb_iv_set()` to set them internally as needed.

eregon replied:
Actually we could even go further and remove the `attr_reader` for `offsets`, like the Java Source object does.
Since offsets is an implementation detail of how Source#line works, and it might change.
But that seems out of scope for this PR.

kddnewton reviewed (changes requested):
Generally fine with the changes, just a couple of things I'd like to see changed.

eregon commented:
@kddnewton I think I addressed all your comments, could you merge?
Notably I have some changes on top of this PR so I need this to be merged to open the PR.
